packages/muelu/utils/matrix/ascii2binary.cpp

```diff
@@ -47,27 +47,23 @@ int main(int argc, char* argv[]) {
 
     ifs >> i >> j >> v;
     i--; j--;
-    while (row != i && ifs.good()) {
-        row = i;
-
-        inds.resize(0);
-        vals.resize(0);
-
-        do {
-            inds.push_back(j);
-            vals.push_back(v);
-
-            ifs >> i >> j >> v;
-            i--; j--;
-        } while (row == i && ifs.good());
-
-
-        int rownnz = inds.size();
-        ofs.write(reinterpret_cast<char*>(&row),    sizeof(row));
-        ofs.write(reinterpret_cast<char*>(&rownnz), sizeof(rownnz));
-        for (int k = 0; k < rownnz; k++) ofs.write(reinterpret_cast<char*>(&inds[0] + k), sizeof(inds[k]));
-        for (int k = 0; k < rownnz; k++) ofs.write(reinterpret_cast<char*>(&vals[0] + k), sizeof(vals[k]));
-
+    for (int row = 0; row < m; row++) {
+      inds.resize(0);
+      vals.resize(0);
+      while (row == i) {
```

Comment:
Does this assume the matrix is stored with rows altogether and in strictly ascending order?

Reply:
Yes, it does. But I don't think that that assumption has changed from before.

Reply:
The original code assumes all the values in a row are contiguously stored, but I think the rows can be in any order.  If I understand the logic of the original code, if a new row is encountered, the check `while (row == i ...` is false, but `row==i+1` isn't necessarily true.

Reply:
Ah, bugger. Yes, that's correct. I guess in that case, the simplest way around all these restrictions is to read the matrix using the Tpetra MM reader, sort it, and then dump to binary.

Reply:
What do you think of this hacky idea.  Keep a boolean array of length num_rows.  As a row is processed, mark its entry true.  After all rows have been processed, for all rows with false entries, add a "-1" entry in the binary file.

Reply:
yup, that'll work.